MaxText/standalone_dataloader.py

```diff
@@ -24,52 +24,183 @@
+            )
+
+            jax.experimental.multihost_utils.sync_global_devices(
+                "Barrier before proceeding to the next step"
+            )
 
+            barrier_end = datetime.datetime.now()
+            max_logging.log(
+                f"STANDALONE DATALOADER : Barrier on host {jax.process_index()} completed on step {step_count} at {barrier_end}, lasted {barrier_end - barrier_start} seconds"
+            )
 
-def data_load_loop(config, state=None):
-  """Main data loader loop.
-  Loads batches of data for each training step.
-  """
-  _, _, _, _, _, _, _, data_iterator, _, state = setup_train_loop(config)
+            step_count += 1
 
-  example_batch = None
+            # Count the per-step time.
+            step_end = datetime.datetime.now()
+            step_time[i].append((step_end - step_start).total_seconds())
+            step_start = step_end
+            # Reset the start time of computing data loading.
+            step_data_loading_start = datetime.datetime.now()
 
-  start = datetime.datetime.now()
-  start_step = get_first_step(state)
-  example_batch = load_next_batch(data_iterator, example_batch, config)
-  jax.block_until_ready(example_batch)
-  first_end = datetime.datetime.now()
-  time_to_load_first_batch = first_end - start
-  if jax.process_index() == 0:
-    max_logging.log(f"STANDALONE DATALOADER : First step completed in {time_to_load_first_batch} seconds, on host 0")
+        epoch_end = datetime.datetime.now()
+        max_logging.log(
+            f"STANDALONE DATALOADER : Host {jax.process_index()} completed epoch {i} using {epoch_end - epoch_start} seconds"
+        )
+        epoch_times.append((epoch_end - epoch_start).total_seconds())
 
-  for _ in np.arange(start_step + 1, config.steps):
-    example_batch = load_next_batch(data_iterator, example_batch, config)
+    training_end = datetime.datetime.now()
+    time_to_load_first_batch = training_end - training_start
 
-  jax.block_until_ready(example_batch)  # wait until the last batch is read
-  end = datetime.datetime.now()
-  if jax.process_index() == 0:
-    max_logging.log(f"STANDALONE DATALOADER : {config.steps} batches loaded in {end-start} seconds, on host 0")
-  return state
+    max_logging.log(
+        f"STANDALONE DATALOADER Metrics: Training completed in {time_to_load_first_batch} seconds, on host {jax.process_index()}"
+    )
+    for i in range(config.epochs):
+        max_logging.log(
+            f"STANDALONE DATALOADER Metrics: Per-epoch total times for epoch {i} on host {jax.process_index()}: {epoch_times[i]}."
+        )
+        max_logging.log(
+            f"STANDALONE DATALOADER Metrics: Per-step data loading times for epoch {i} on host {jax.process_index()}: {data_loading_time[i]}."
+        )
+        max_logging.log(
+            f"STANDALONE DATALOADER Metrics: Per-step times for epoch {i} on host {jax.process_index()}: {step_time[i]}."
+        )
 
 
 def main(argv: Sequence[str]) -> None:
```

Comment:
Why the reformatting below here? Is there a specific formatter that MaxText is using to make it consistent with the rest of this codebase? Same for other files in this PR

Reply:
Ack. The default Python formatter we use gives 4 spaces for tabs and we need to use 2 spaces. Also [recommended](go/pystyle#indentation) by internal Python style guide.

Reply:
Modified the format to align with 2 spaces for indentation.